File: src-python/planarvibe/layouts/tutte.py
"""Tutte barycentric layout. Literal port of static/js/layout-tutte.js.

Entry point: `apply_layout(graph, initial_positions=None, options=None)`.
Mirrors the JS PlanarVibeTutte.applyLayout but without the Cytoscape runtime;
preprocessing chain is driven directly via `preprocessing.prepare_graph_data`.
"""

from __future__ import annotations

import math
from typing import Sequence

from .. import geometry as geo
from .. import graph as gu
from .. import linear_algebra as la
from .. import preprocessing


def build_tutte_weights(graph, augmented_graph) -> dict[str, float]:
    original_pairs = graph.edge_pairs
    augmented_pairs = augmented_graph.edge_pairs
    adjacency = augmented_graph.adjacency
    degree_by_id: dict[str, int] = {}
    original_edge_set: dict[str, bool] = {}
    outer_dummy_ids = getattr(augmented_graph, "_outer_dummy_ids", None) or []
    outer_dummy_set = {str(x) for x in outer_dummy_ids}
    weights: dict[str, float] = {}

    for (u, v) in original_pairs:
        original_edge_set[gu.edge_key(u, v)] = True

    use_adjacency = adjacency is not None and len(adjacency) > 0
    if not use_adjacency:
        for (u, v) in augmented_pairs:
            degree_by_id[u] = degree_by_id.get(u, 0) + 1
            degree_by_id[v] = degree_by_id.get(v, 0) + 1

    for (u, v) in augmented_pairs:
        key = gu.edge_key(u, v)
        touches_outer_dummy = str(u) in outer_dummy_set or str(v) in outer_dummy_set
        base_weight = 10.0 if (key not in original_edge_set and touches_outer_dummy) else 1.0
        if use_adjacency:
            du = max(1, len(adjacency.get(u, [])))
            dv = max(1, len(adjacency.get(v, [])))
        else:
            du = max(1, degree_by_id.get(u, 0))
            dv = max(1, degree_by_id.get(v, 0))
        base_weight /= math.sqrt(du * dv)
        weights[key] = base_weight
    return weights


def default_outer_placement_options(overrides: dict | None = None) -> dict:
    base = {
        "defaultCenterX": 450,
        "defaultCenterY": 310,
        "defaultRadius": 300,
        "outerRotation": None,
    }
    if overrides:
        base.update(overrides)
    return base


def place_outer_face_vertices(node_ids: Sequence[str], outer_face: Sequence[str], options: dict | None = None) -> dict:
    ids = [str(x) for x in (node_ids or [])]
    face = gu.normalize_outer_face(outer_face)
    opts = default_outer_placement_options(options)
    pos: dict[str, tuple[float, float]] = {nid: (0.0, 0.0) for nid in ids}

    cx = opts["defaultCenterX"]
    cy = opts["defaultCenterY"]
    R = opts["defaultRadius"]

    if not face:
        return pos

    gamma = 2 * math.pi / len(face)
    if opts.get("outerRotation") is not None and math.isfinite(float(opts["outerRotation"])):
        start_angle = float(opts["outerRotation"])
    else:
        start_angle = math.pi / 2 - gamma / 2

    for i, v in enumerate(face):
        v = str(v)
        angle = start_angle + gamma * i
        pos[v] = (cx + R * math.cos(angle), cy + R * math.sin(angle))

    fixed = opts.get("fixedOuterPos")
    if isinstance(fixed, dict):
        for v in face:
            fp = fixed.get(str(v))
            if fp is not None and geo._is_finite(fp[0]) and geo._is_finite(fp[1]):
                pos[str(v)] = (fp[0], fp[1])

    return pos


def _barycentric_error(message: str, graph, face) -> dict:
    return gu.build_layout_error({
        "message": message,
        "graph": graph,
        "outerFace": face,
    })


def compute_barycentric_positions(graph, outer_face: Sequence[str], options: dict) -> dict:
    ids = graph.node_ids
    face = gu.normalize_outer_face(outer_face)
    adjacency = options.get("adjacency") or graph.adjacency
    weights = options.get("weights")
    init_options = options.get("initOptions") or default_outer_placement_options()
    pos = place_outer_face_vertices(ids, face, init_options)
    outer_set = set(face)
    interior_ids: list[str] = []
    interior_index_by_id: dict[str, int] = {}

    if len(ids) < 1:
        return _barycentric_error("No vertices", graph, face)
    if len(face) < 3:
        return _barycentric_error("Outer face is invalid", graph, face)
    if not isinstance(weights, dict):
        return _barycentric_error("Barycentric weights are required", graph, face)

    for nid in ids:
        if nid not in outer_set:
            interior_index_by_id[nid] = len(interior_ids)
            interior_ids.append(nid)
    if not interior_ids:
        return gu.build_layout_result({
            "graph": graph,
            "outerFace": face,
            "positions": pos,
            "iters": 1,
        })

    n = len(interior_ids)
    L = [[0.0] * n for _ in range(n)]
    bx = [0.0] * n
    by = [0.0] * n
    for i, v in enumerate(interior_ids):
        L[i][i] = 1.0
        neighbors = adjacency.get(v, [])
        raw_weights: list[float] = [0.0] * len(neighbors)
        weight_sum = 0.0
        for j, u_raw in enumerate(neighbors):
            u = str(u_raw)
            w = weights.get(gu.edge_key(v, u))
            if not isinstance(w, (int, float)) or not math.isfinite(w) or w <= 0:
                w = 1.0
            raw_weights[j] = w
            weight_sum += w
        if not (weight_sum > 0):
            continue
        for j, u_raw in enumerate(neighbors):
            u = str(u_raw)
            w = raw_weights[j] / weight_sum
            interior_idx = interior_index_by_id.get(u)
            if interior_idx is None:
                bx[i] += w * pos[u][0]
                by[i] += w * pos[u][1]
            else:
                L[i][interior_idx] -= w

    factor = la.lu_factorize(L)
    solved = la.solve_lu_with_two_rhs(factor, bx, by) if factor else None
    if solved is None:
        return _barycentric_error("Exact barycentric solve failed", graph, face)
    x1, x2 = solved
    for i, nid in enumerate(interior_ids):
        pos[nid] = (x1[i], x2[i])

    return gu.build_layout_result({
        "graph": graph,
        "outerFace": face,
        "positions": pos,
        "iters": 1,
    })


def _compute_tutte_layout_with_prepared(prepared: dict) -> dict:
    if not prepared or not prepared.get("ok"):
        return gu.build_layout_error(prepared or {"message": "Tutte failed"})

    graph = prepared["graph"]
    ids = graph.node_ids
    pairs = graph.edge_pairs

    augmented_graph = prepared["augmented"]["graph"]
    augmented_outer = prepared["augmentedOuterFace"]
    barycentric = compute_barycentric_positions(
        augmented_graph,
        augmented_outer,
        {
            "initOptions": default_outer_placement_options(),
            "weights": build_tutte_weights(prepared["graph"], augmented_graph),
        },
    )
    if not barycentric or not barycentric.get("ok") or not barycentric.get("positions"):
        return gu.build_layout_error(barycentric or {"message": "Tutte failed"})

    projected = geo.filter_position_map(barycentric["positions"], ids)
    has_crossings = geo.has_position_crossings(projected, pairs)
    if has_crossings:
        return gu.build_layout_error({
            "message": "Tutte produced a non-plane drawing",
            "graph": prepared["graph"],
            "outerFace": prepared["outerFace"],
            "augmented": prepared["augmented"],
        })

    return gu.build_layout_result({
        "nodeIds": ids,
        "edgePairs": pairs,
        "outerFace": prepared["outerFace"],
        "embedding": prepared["baseEmbedding"],
        "augmented": prepared["augmented"],
        "graph": prepared["graph"],
        "debugPositions": barycentric["positions"],
        "positions": projected,
        "iters": barycentric.get("iters"),
    })


def apply_layout(graph, initial_positions: dict | None = None, options: dict | None = None) -> dict:
    opts = options or {}
    prepared = preprocessing.prepare_graph_data(graph, {
        "failureLabel": "Tutte layout",
        "augmentationMethod": opts.get("augmentationMethod"),
        "augmentationOptions": opts.get("augmentationOptions"),
        "currentPositions": initial_positions or {},
    })
    if not prepared or not prepared.get("ok"):
        return gu.build_layout_error(prepared or {"message": "Tutte failed"})
    result = _compute_tutte_layout_with_prepared(prepared)
    if result.get("ok"):
        outer = result.get("outerFace") or []
        result["message"] = gu.build_layout_status_message("Tutte", {
            "outerFaceVertexCount": len(outer),
            "iters": result.get("iters"),
        })
    return result

File: src-cpp/src/tutte.cpp
#include "layouts/tutte.hpp"

#include "geometry.hpp"
#include "linear_algebra.hpp"
#include "preprocessing.hpp"

#include <algorithm>
#include <cmath>
#include <unordered_map>
#include <unordered_set>

namespace planarvibe::layouts {

namespace {

std::string ekey(const std::string& a, const std::string& b) {
    return (a < b) ? (a + "::" + b) : (b + "::" + a);
}

} // namespace

pg::PosByStr place_outer_face_vertices(
    const std::vector<std::string>& node_ids,
    const std::vector<std::string>& outer_face,
    const TutteOuterPlacement& opts) {
    pg::PosByStr pos;
    for (const auto& nid : node_ids) pos[nid] = {0.0, 0.0};
    if (outer_face.empty()) return pos;
    double gamma = 2.0 * M_PI / outer_face.size();
    double start_angle = opts.outer_rotation ? *opts.outer_rotation : (M_PI / 2.0 - gamma / 2.0);
    for (size_t i = 0; i < outer_face.size(); ++i) {
        double a = start_angle + gamma * i;
        pos[outer_face[i]] = {opts.cx + opts.R * std::cos(a), opts.cy + opts.R * std::sin(a)};
    }
    if (opts.fixed_outer_pos) {
        for (const auto& nid : outer_face) {
            auto it = opts.fixed_outer_pos->find(nid);
            if (it == opts.fixed_outer_pos->end()) continue;
            double x = it->second[0], y = it->second[1];
            if (std::isfinite(x) && std::isfinite(y)) pos[nid] = {x, y};
        }
    }
    return pos;
}

std::unordered_map<std::string, double> build_tutte_weights(
    const std::vector<std::pair<std::string,std::string>>& original_pairs,
    const std::vector<std::pair<std::string,std::string>>& augmented_pairs,
    const std::vector<std::string>& outer_dummy_ids) {
    std::unordered_map<std::string, double> weights;
    std::unordered_set<std::string> outer_dummy_set(outer_dummy_ids.begin(), outer_dummy_ids.end());
    std::unordered_set<std::string> original_edge_set;
    for (const auto& [u, v] : original_pairs) original_edge_set.insert(ekey(u, v));

    std::unordered_map<std::string, int> deg;
    for (const auto& [u, v] : augmented_pairs) { deg[u]++; deg[v]++; }

    for (const auto& [u, v] : augmented_pairs) {
        std::string k = ekey(u, v);
        bool touches_outer = outer_dummy_set.count(u) || outer_dummy_set.count(v);
        double base = (!original_edge_set.count(k) && touches_outer) ? 10.0 : 1.0;
        int du = std::max(1, deg.count(u) ? deg[u] : 0);
        int dv = std::max(1, deg.count(v) ? deg[v] : 0);
        base /= std::sqrt(double(du) * double(dv));
        weights[k] = base;
    }
    return weights;
}

BarycentricResult compute_barycentric_positions(
    const std::vector<std::string>& node_ids,
    const std::unordered_map<std::string, std::vector<std::string>>& adjacency,
    const std::vector<std::string>& outer_face,
    const std::unordered_map<std::string, double>& weights,
    const TutteOuterPlacement& placement) {
    BarycentricResult r;
    if (node_ids.empty()) { r.message = "No vertices"; return r; }
    if (outer_face.size() < 3) { r.message = "Outer face is invalid"; return r; }

    auto pos = place_outer_face_vertices(node_ids, outer_face, placement);
    std::unordered_set<std::string> outer_set(outer_face.begin(), outer_face.end());
    std::vector<std::string> interior;
    std::unordered_map<std::string, int> interior_idx;
    for (const auto& nid : node_ids) {
        if (outer_set.count(nid)) continue;
        interior_idx[nid] = (int)interior.size();
        interior.push_back(nid);
    }
    if (interior.empty()) {
        r.ok = true;
        r.iters = 1;
        r.positions = pos;
        return r;
    }

    int n = (int)interior.size();
    std::vector<std::vector<double>> L(n, std::vector<double>(n, 0.0));
    std::vector<double> bx(n, 0.0), by(n, 0.0);
    for (int i = 0; i < n; ++i) {
        L[i][i] = 1.0;
        auto it = adjacency.find(interior[i]);
        if (it == adjacency.end()) continue;
        const auto& neighbors = it->second;
        std::vector<double> raw(neighbors.size(), 0.0);
        double sum = 0.0;
        for (size_t j = 0; j < neighbors.size(); ++j) {
            auto wit = weights.find(ekey(interior[i], neighbors[j]));
            double w = (wit != weights.end() && std::isfinite(wit->second) && wit->second > 0) ? wit->second : 1.0;
            raw[j] = w;
            sum += w;
        }
        if (!(sum > 0)) continue;
        for (size_t j = 0; j < neighbors.size(); ++j) {
            double w = raw[j] / sum;
            auto iit = interior_idx.find(neighbors[j]);
            if (iit == interior_idx.end()) {
                auto pit = pos.find(neighbors[j]);
                if (pit == pos.end()) continue;
                bx[i] += w * pit->second[0];
                by[i] += w * pit->second[1];
            } else {
                L[i][iit->second] -= w;
            }
        }
    }

    auto factor = la::lu_factorize(L);
    if (!factor) { r.message = "Exact barycentric solve failed"; return r; }
    auto solved = la::solve_lu_with_two_rhs(*factor, bx, by);
    if (!solved) { r.message = "Exact barycentric solve failed"; return r; }
    for (int i = 0; i < n; ++i) pos[interior[i]] = {solved->first[i], solved->second[i]};

    r.ok = true;
    r.iters = 1;
    r.positions = std::move(pos);
    return r;
}

LayoutResult tutte(const Graph& g, const pg::PosByStr* initial_positions) {
    LayoutResult r;
    preprocessing::PrepareConfig cfg;
    cfg.failure_label = "Tutte layout";
    cfg.current_positions = initial_positions;
    auto prep = preprocessing::prepare_graph_data(g, cfg);
    if (!prep.ok) {
        r.ok = false;
        r.message = prep.message.empty() ? "Tutte failed" : prep.message;
        return r;
    }

    // Build adjacency for augmented graph.
    std::unordered_map<std::string, std::vector<std::string>> adjacency;
    for (const auto& nid : prep.augmented_node_ids) adjacency[nid] = {};
    for (const auto& [u, v] : prep.augmented_edge_pairs) {
        adjacency[u].push_back(v);
        adjacency[v].push_back(u);
    }

    auto weights = build_tutte_weights(prep.edge_pairs, prep.augmented_edge_pairs, prep.outer_dummy_ids);

    auto br = compute_barycentric_positions(prep.augmented_node_ids, adjacency,
                                            prep.augmented_outer_face, weights);
    if (!br.ok) {
        r.ok = false;
        r.message = br.message.empty() ? "Tutte failed" : br.message;
        return r;
    }

    r.iters = br.iters;
    r.positions.resize(g.n);
    for (int i = 0; i < g.n; ++i) {
        auto it = br.positions.find(g.node_names[i]);
        if (it != br.positions.end() && std::isfinite(it->second[0]) && std::isfinite(it->second[1])) {
            r.positions.put(i, it->second[0], it->second[1]);
        }
    }
    if (geo::has_position_crossings(r.positions, g.edges)) {
        r.ok = false;
        r.message = "Tutte produced a non-plane drawing";
    } else {
        r.ok = true;
        r.message = "Tutte layout";
    }
    return r;
}

} // namespace planarvibe::layouts
